frontend-react/src/utils/DiffCompare/JsonDiffer.ts

```diff
@@ -0,0 +1,330 @@
+        (acc: string, each: MarkerInsert) =>
+            `${acc.slice(0, each.pos)}${each.mark}${acc.slice(each.pos)}`,
+        jsonStr
+    );
+};
+
+/**
+ * Diff compares two jsons and returns a map of the differences. You probably want to use jsonDifferMarkup
+ * @param leftData
+ * @param rightData
+ * @return JsonDiffResult Called by jsonDifferMarkup.
+ */
+const jsonDiffer = (
+    leftData: SourceMapResult,
+    rightData: SourceMapResult
+): JsonDiffResult => {
+    // diff the keys. If the key is different, then just consider the value of that key to be different.
+    const leftKeys = Object.keys(leftData.pointers);
+    const rightKeys = Object.keys(rightData.pointers);
+
+    // this is looking for diffs between the two lists.
+    let addedLeftKeys = leftKeys.filter(
+        (key) => key.length && !rightKeys.includes(key)
+    );
+    let addedRightKeys = rightKeys.filter(
+        (key) => key.length && !leftKeys.includes(key)
+    );
+
+    // now we want intersection (aka NOT changed and see if the values have changed).
+    let intersectionKeys = leftKeys.filter((key) => rightKeys.includes(key));
+
+    // for readability improvements only, pull out start/end values
+    const getStartEnd = (key: string, data: SourceMapResult) => {
+        return [data.pointers[key].value.pos, data.pointers[key].valueEnd.pos];
+    };
+
+    // inline getValue() improves readability only. slices out the string of a given value
+    const getValue = (key: string, data: SourceMapResult): string =>
+        data.json.slice(...getStartEnd(key, data));
+
+    // we pull out the value of unchanged keys and see if that's different.
+    let changedKeys = intersectionKeys.filter(
+        (key) =>
+            key !== "" && getValue(key, leftData) !== getValue(key, rightData)
+    );
+
+    // now extract just the node leaves from the keys. This is because technically the content of each parent
+    // has changed, but we really only think about the leaf nodes as being diff
+    changedKeys = extractLeafNodes(changedKeys);
+
+    return {
+        addedLeftKeys,
+        addedRightKeys,
+        changedKeys,
+    };
+};
+
+/**
+ * Given two json find the difference, then return a normalized json string and one with <mark>s
+ * added for the differences.
+ *
+ * @param leftJson Valid JSON.
+ * @param rightJson Valid JSON.
+ */
+export const jsonDifferMarkup = (
+    leftJson: unknown,
+    rightJson: unknown
+): DifferMarkupResult => {
+    const leftMap = jsonSourceMap(leftJson, 2);
+    const rightMap = jsonSourceMap(rightJson, 2);
+    const diffs = jsonDiffer(leftMap, rightMap);
+
+    // so `{level1: { level2: { level3: "value"} } }` vs `{level1: { level2: { level3MOD: "value"} } }`
+    // The keys `/level/level2/level3` (left) and  `/level1/level2/level3MOD` (right) are different,
+    // BUT so is the PARENT VALUE `/level1/level2`
+    // so we go back and remove changedKeys from the addedKeys
+    let leftChanged = diffs.changedKeys;
```

Comment:
I did a bit of normalization around the left/right repeated logic until at first, but I undid it because I felt it made following the logic harder (I think I got it down to 5 actual lines per side).